python/private/interpreter.bzl

```diff
@@ -0,0 +1,87 @@
+# Copyright 2025 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implementation of the rules to access the underlying Python interpreter."""
+
+load("@bazel_skylib//lib:paths.bzl", "paths")
+load("//python:py_runtime_info.bzl", "PyRuntimeInfo")
+load(":sentinel.bzl", "SentinelInfo")
+load(":toolchain_types.bzl", "TARGET_TOOLCHAIN_TYPE")
+load(":py_executable.bzl", "relative_path", "runfiles_root_path")
+
+def _interpreter_binary_impl(ctx):
+    if SentinelInfo in ctx.attr.binary:
+        toolchain = ctx.toolchains[TARGET_TOOLCHAIN_TYPE]
+        runtime = toolchain.py3_runtime
+    else:
+        runtime = ctx.attr.binary[PyRuntimeInfo]
+
+    # NOTE: We name the output filename after the underlying file name
+    # because of things like pyenv: they use $0 to determine what to
+    # re-exec. If it's not a recognized name, then they fail.
+    if runtime.interpreter:
+        # Option 1:
+        # Works locally, but not remotely.
+        #executable = ctx.actions.declare_file(runtime.interpreter.basename)
+        #ctx.actions.symlink(output = executable, target_file = runtime.interpreter)
+
+        # Option 2:
+        # Works locally, and remotely, but not on Windows.
+        executable = ctx.actions.declare_file(runtime.interpreter.basename)
+        ctx.actions.expand_template(
+            template = ctx.file._template,
+            output = executable,
+            substitutions = {
+                "%target_file%": runtime.interpreter.short_path,
```

Comment:
I think this might be missing the workspace name? runfiles paths are of the form `<repo>/<runfiles path within repo>`.

Or, that's the most sensible way I've started using with them. There might be some special casing in runfiles lookup libraries. I've found those to be confusing, though; less magic is involved when the string directly matches what you see in the file system.

There's a helper in py_executable to compute the runfiles path for a given file and ctx; ah yeah, i see that mentioned in the other comment.

Reply:
Ah, I see -- the shell code is handling the ../ part.

I've updated the template expansion to pass a runfiles-root relative path and removed the post-processing in the shell script.